content/posts/exactly-once-http.md

```diff
@@ -0,0 +1,177 @@
+
+#### Phase 1 - PUT
+
+The first phase is using `PUT` to transmit the request payload created by the user code to the remote service. This is done immediately in the message handling code in the same way as creating a blob works (described in [the previous post](https://exactly-once.github.io/posts/side-effects/)). The cleanup mechanism ensures that requests that are associated with abandoned message processing attempts are removed.
+
+#### Phase 2 - POST
+
+The second phase happens during the side effects publication. In this phase the caller issues `POST` to the same URL as previous `PUT`. This is the moment when the target service actually does the processing. The response is not returned directly but stored by the target services in a blob.
+
+#### Phase 3 - GET
+
+The third phase happens before the _follow-up message_ is passed to the message handler. The infrastructure issues `GET` to obtain the response generated in the previous phase. That response is added to the message handling context.
+
+#### Phase 4 - DELETE
+
+The fourth phase of the HTTP interaction happens during the side effects publication stage of processing the _follow-up message_. A `DELETE` is issued removes the response information from the service, making sure that the whole interaction leaves no garbage.
+
+### Deduplication
+
+All good but where is the deduplication? You are right, we have just defined how the HTTP interaction looks like in the context of message processing but have not done anything to deduplicate the calls. Well, actually we did. Let's discuss this by analyzing what can happen in each phase if things go wrong.
+
+If the caller fails after the `PUT` phase and the message is retried, another request is generated and transmitted via `PUT`. At the end of the successful attempt all abandoned requests are removed via `DELETE`. No problem with duplication here.
+
+If the caller fails after the `POST` phase the same `POST` will be issued again because it has been persisted as a side effect awaiting publication. This means that the server can receive multiple `POST` calls for a single request transmitted previously by `PUT`. How can we deal with that? Careful readers can see an analogy to token-based-deduplication here. The request stored as a blob in the target service is an equivalent to a token. It serves as the deduplication state. The implementation of the service marks the request as processed when it executes the `POST` so that subsequent `POST`s are considered duplicates and ignored. In addition to marking the request as processed, the `POST` also stores the response in a dedicated blob so that it can be retrieved by the caller via `GET`.
```

Comment:
```suggestion
If the caller fails after the `POST` phase the same `POST` will be issued again because it has been persisted as a side effect awaiting publication. This means that the server can receive multiple `POST` calls for a single request transmitted previously by `PUT`. How can we deal with that? Careful readers can see an analogy to token-based-deduplication here. The request stored as a blob in the target service is equivalent to a token. It serves as the deduplication state. The implementation of the service marks the request as processed when it executes the `POST` so that subsequent `POST`s are considered duplicates and ignored. In addition to marking the request as processed, the `POST` also stores the response in a dedicated blob so that it can be retrieved by the caller via `GET`.
```